crates/sparrow-scheduler/src/worker.rs

```diff
@@ -35,49 +38,89 @@ pub trait Scheduler {
 #[derive(Debug, Clone)]
 pub struct Injector {
     queue: GlobalQueue<TaskRef>,
+    pub(crate) idle_workers: Arc<IdleWorkers>,
 }
 
 impl Injector {
     pub fn create(workers: usize, local_queue_size: u16) -> (Self, Vec<Worker>) {
+        let idle_workers = IdleWorkers::new(workers);
         let queue = GlobalQueue::new(workers, local_queue_size);
         let workers = queue
             .take_local_queues()
-            .map(|queue| Worker { queue })
+            .map(|queue| Worker {
+                queue,
+                idle_workers: idle_workers.clone(),
+            })
             .collect();
-        (Injector { queue }, workers)
+        (
+            Injector {
+                queue,
+                idle_workers,
+            },
+            workers,
+        )
     }
 }
 
 impl Scheduler for Injector {
     fn schedule_global(&self, task: TaskRef) {
         if task.schedule() {
-            self.queue.push(task)
+            tracing::trace!("1 Added {task:?} to queue");
+            self.queue.push(task);
+        } else {
+            tracing::trace!("1 {task:?} already executing");
         }
+        self.idle_workers.wake_one();
     }
 
     fn schedule(&mut self, task: TaskRef) {
-        self.schedule_global(task)
+        self.schedule_global(task);
     }
 
     fn schedule_yield(&mut self, task: TaskRef) {
-        self.schedule_global(task)
+        self.schedule_global(task);
     }
 }
 
 /// An individual worker that allows adding work to the local or global queue.
 pub struct Worker {
     queue: LocalQueue<TaskRef>,
+    idle_workers: Arc<IdleWorkers>,
 }
 
 impl Worker {
     /// Run the work loop to completion.
-    pub(crate) fn work_loop(mut self) -> error_stack::Result<(), Error> {
-        while let Some(task) = self.queue.pop() {
-            if task.do_work(&mut self)? {
-                // This means that the task was schedule while we were executing.
-                // As a result, we didn't add it to any queue yet, so we need to
-                // do so now.
-                self.queue.push_global(task);
+    pub(crate) fn work_loop(mut self, index: usize) -> error_stack::Result<(), Error> {
+        let thread_id = std::thread::current().id();
+        let _span = tracing::info_span!("Worker", ?thread_id, index).entered();
+        tracing::info!("Starting work loop on thread {thread_id:?}");
```

Comment:
General question: If we're within the span here, do we need to reiterate variables that are already apart of the span? Does having `thread_id` in this `info` duplicate that variable in the log line?

Reply:
We don't need to repeat them, although duplication depends on the tracing backend. Eg. in jaeger you may see the span as a section in the visualization, while the log line is a specific point. On the console, we can configure whether the span is included at all, etc.

Generally -- doesn't seem to *hurt* to have it repeated, although as you say, probably not necessary. I'll drop it.